src/main/java/org/broadinstitute/hellbender/utils/read/ArtificialBAMBuilder.java

```diff
@@ -0,0 +1,222 @@
+            final String sample = "sample" + i;
+            samples.add(sample);
+            rg.setSample(sample);
+            rg.setPlatform(NGSPlatform.ILLUMINA.getDefaultPlatform());
+            header.addReadGroup(rg);
+        }
+
+        return this;
+    }
+
+    public List<String> getSamples() {
+        return samples;
+    }
+
+    /**
+     * Create a read stream based on the parameters.  The cigar string for each
+     * read will be *M, where * is the length of the read.
+     *
+     * Useful for testing things like LocusIteratorBystate
+     *
+     * @return a ordered list of reads
+     */
+    public List<GATKRead> makeReads() {
+        if ( createdReads == null ) {
+            final String baseName = "read";
+            final LinkedList<SAMReadGroupRecord> readGroups = new LinkedList<>();
+            for ( final SAMReadGroupRecord rg : header.getReadGroups()) {
+                readGroups.add(rg);
+            }
+
+            final List<GATKRead> reads = new ArrayList<>(nReadsPerLocus*nLoci);
+            for ( int locusI = 0; locusI < nLoci; locusI++) {
+                final int locus = locusI * (skipNLoci + 1);
+                for ( int readI = 0; readI < nReadsPerLocus; readI++ ) {
+                    for ( final SAMReadGroupRecord rg : readGroups ) {
+                        final String readName = String.format("%s.%d.%d.%s", baseName, locus, readI, rg.getId());
+                        final GATKRead read = ArtificialReadUtils.createArtificialRead(header, readName, 0, alignmentStart + locus, readLength);
+                        read.setReadGroup(rg.getId());
+                        reads.add(read);
+                    }
+                }
+            }
+
+            if ( ! additionalReads.isEmpty() ) {
+                reads.addAll(additionalReads);
+                Collections.sort(reads, new ReadCoordinateComparator(header));
+            }
+
+            createdReads = new ArrayList<>(reads);
+        }
+
+        return createdReads;
+    }
+
+    /**
+     * Make an indexed BAM file contains the reads in the builder, marking it for deleteOnExit()
+     * @return the BAM file
+     */
+    public File makeTemporarilyBAMFile() throws IOException{
+        final File file = File.createTempFile("tempBAM", ".bam");
+        file.deleteOnExit();
+
+        // Register the bam index file for deletion on exit as well:
+        new File(file.getAbsolutePath().replace(".bam", ".bai")).deleteOnExit();
+        new File(file.getAbsolutePath() + ".bai").deleteOnExit();
+
+        return makeBAMFile(file);
+    }
+
+    /**
+     * Write the reads from this builder to output, creating an index as well
+     * @param output the output BAM file we want to use
+     */
+    public File makeBAMFile(final File output) {
+        Utils.nonNull(output);
+        try(final SAMFileWriter writer = factory.makeBAMWriter(header, true, output, 0)) {
```

Comment:
Set `preSorted` to false on the writer to force the writer to sort the reads properly (for the reason given in a comment above).